### PR TITLE
`bot/build.sh` using `eessi_container.sh`

### PR DESCRIPTION
**Pull request by trz42**
First version of the script to be called by the EESSI bot. Works with the updated script `eessi_container.sh` (see #232).

First version of `bot/build.sh`
* `.json` & `yq` have been replaced by a `.cfg` (`.ini` file format) & `scripts/cfg_files.sh`
  * ~requires tool `yq` to parse `.json` files containing information from the bot to the `bot/build.sh`~
  * ~`yq` is required where the bot jobs run (since it is a binary versions for `x86_64` and `aarch64` have to be installed)~
  * ~while `yq` seems to be a nice tool, we may not want to depend on it and also not want to deal with different binaries; a possible change might require to not handover information in a `.json` file but rather a simpler `.ini` or `.cfg` file and then use parsing functions provided by `scripts/cfg_files.sh`~
* also adds passing `--generic` to `install_software_layer.sh`

Updates to `eessi_container.sh`
* PR also contains small update to `eessi_container.sh` itself (improving output messages concerning `--resume` information)
* adds support for a command line argument flag terminator `--`, e.g., can be called
  ```
  ./eessi_container.sh ARGS_TO_eessi_container.sh -- other_script_to_run_in_container.sh ARGS_TO_other_script
  ```

Bot-side code is added through https://github.com/EESSI/eessi-bot-software-layer/pull/155

### CODE REVIEW

**Comment by boegel** [comment]
@trz42 Please sync with `main` now that #232 is merged...

File: eessi_container.sh [hunk 1]
```diff
@@ -173,6 +173,11 @@ while [[ $# -gt 0 ]]; do
       export https_proxy=${HTTPS_PROXY}
       shift 2
       ;;
+    --)
```
Comment (boegel) [inline eessi_container.sh:176]: We should make sure the use of `--` is also reflected in the docs at https://eessi.github.io/docs/getting_access/eessi_container/#running-a-simple-command

File: bot/build.sh [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    GENERIC_OPT="--generic"
+fi
+
+mkdir -p previous_tmp/{build_step,tarball_step}
+build_outerr=$(mktemp build.outerr.XXXX)
+echo "Executing command to build software:"
+echo "./eessi_container.sh ${BUILD_STEP_ARGS[@]}"
+echo "                     -- ./install_software_layer.sh ${GENERIC_OPT} \"$@\" 2>&1 | tee -a ${build_outerr}"
+# set EESSI_REPOS_CFG_DIR_OVERRIDE to ./cfg
+export EESSI_REPOS_CFG_DIR_OVERRIDE=${PWD}/cfg
```
Comment (boegel) [inline bot/build.sh:173]: This should be removed, already done above

File: bot/build.sh [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+# license: GPLv2
+#
+
+# ASSUMPTIONs:
+#  - working directory has been prepared by the bot with a checkout of a
+#    pull request (OR by some other means)
+#  - the working directory contains a directory 'cfg' where the main config
+#    file 'job.cfg' has been deposited
+#  - the directory may contain any additional files references in job.cfg
+
```
Comment (boegel) [inline bot/build.sh:20]: let's add a `set -e` here to stop the script as soon as something failed?

File: bot/build.sh [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+#     way to determine the IP address of a proxy)
+#   - bot needs to make repos.cfg and cfg_bundle available to job (likely, by copying
+#     files into './cfg/.' and defining '.repository.repos_cfg_dir' in './cfg/job.cfg')
+
+# prepare options and directories for calling eessi_container.sh
+declare -a BUILD_STEP_ARGS=()
+BUILD_STEP_ARGS+=("--verbose")
+BUILD_STEP_ARGS+=("--access" "rw")
+BUILD_STEP_ARGS+=("--mode" "run")
+BUILD_STEP_ARGS+=("--save" "${PWD}/previous_tmp/build_step")
```
Comment (boegel) [inline bot/build.sh:131]: introduce a `${PREVIOUS_TMP_DIR}` variable for `${PWD}/previous_tmp`?

File: bot/build.sh [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+echo "./eessi_container.sh ${TARBALL_STEP_ARGS[@]}"
+echo "                     -- ./create_tarball.sh ${TMP_IN_CONTAINER} ${EESSI_PILOT_VERSION} ${EESSI_SOFTWARE_SUBDIR_OVERRIDE} /eessi_bot_job/${TGZ} 2>&1 | tee -a ${tar_outerr}"
+./eessi_container.sh "${TARBALL_STEP_ARGS[@]}" \
+                     -- ./create_tarball.sh ${TMP_IN_CONTAINER} ${EESSI_PILOT_VERSION} ${EESSI_SOFTWARE_SUBDIR_OVERRIDE} /eessi_bot_job/${TGZ} 2>&1 | tee -a ${tar_outerr}
+
+# if two tarballs have been generated, only keep the one from tarball step
+NUM_TARBALLS=$(find ${PWD}/previous_tmp -type f -name "*tgz" | wc -l)
+if [[ ${NUM_TARBALLS} -eq 2 ]]; then
+    rm -f previous_tmp/build_step/*.tgz
+fi
```
Comment (boegel) [inline bot/build.sh:202]: This looks very specific, maybe comment out for now (and make sure there's an issue for this)

**Review by boegel — CHANGES_REQUESTED** [review]

File: bot/build.sh (hunk 4 above)
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh (hunk 3 above)
Comment (trz42) [reply bot/build.sh:20]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh (hunk 2 above)
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+BUILD_STEP_ARGS+=("--save" "${PWD}/previous_tmp/build_step")
+BUILD_STEP_ARGS+=("--storage" "${STORAGE}")
+declare -a TARBALL_STEP_ARGS=()
+TARBALL_STEP_ARGS+=("--verbose")
+TARBALL_STEP_ARGS+=("--access" "rw")
+TARBALL_STEP_ARGS+=("--mode" "run")
+TARBALL_STEP_ARGS+=("--save" "${PWD}/previous_tmp/tarball_step")
+CONTAINER_OPT=
+if [[ ! -z ${CONTAINER} ]]; then
+    CONTAINER_OPT="--container ${CONTAINER}"
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+# TODO
+#   - CODED add handling of EESSI_SOFTWARE_SUBDIR_OVERRIDE to eessi_container.sh
+#     TODO ensure that the bot makes use of that. (currently sets env var
+#     CPU_TARGET & adds --export=ALL,CPU_TARGET=val to sbatch command ... also
+#     add it to cfg/job.cfg - .architecture.software_subdir)
+#   - CODED add handling of http(s)_proxy to eessi_container.sh, in there needs the
+#     CVMFS_HTTP_PROXY added to /etc/cvmfs/default.local (this needs a robust
+#     way to determine the IP address of a proxy)
+#   - bot needs to make repos.cfg and cfg_bundle available to job (likely, by copying
+#     files into './cfg/.' and defining '.repository.repos_cfg_dir' in './cfg/job.cfg')
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+# determine CVMFS repo to be used from .repository.repo_name in cfg/job.cfg
+# here, just set EESSI_CVMFS_REPO_OVERRIDE, a bit further down
+# "source init/eessi_defaults" via sourcing init/minimal_eessi_env
+export EESSI_CVMFS_REPO_OVERRIDE=$(cfg_get_value "repository" "repo_name")
+
+
+# determine architecture to be used from entry .architecture in cfg/job.cfg
+# default: leave empty to let downstream script(s) determine subdir to be used
+if [[ ! -z "${CPU_TARGET}" ]]; then
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+# determine pilot version to be used from .repository.repo_version in cfg/job.cfg
+# here, just set & export EESSI_PILOT_VERSION_OVERRIDE
+# next script (eessi_container.sh) makes use of it via sourcing init scripts
+# (e.g., init/eessi_defaults or init/minimal_eessi_env)
+export EESSI_PILOT_VERSION_OVERRIDE=$(cfg_get_value "repository" "repo_version")
+
+# determine CVMFS repo to be used from .repository.repo_name in cfg/job.cfg
+# here, just set EESSI_CVMFS_REPO_OVERRIDE, a bit further down
+# "source init/eessi_defaults" via sourcing init/minimal_eessi_env
+export EESSI_CVMFS_REPO_OVERRIDE=$(cfg_get_value "repository" "repo_name")
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh:111]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        module load ${mod}
+    done
+else
+    echo "bot/build.sh: no modules to be loaded"
+fi
+
+# determine repository to be used from entry .repository in cfg/job.cfg
+REPOSITORY=$(cfg_get_value "repository" "repo_id")
+EESSI_REPOS_CFG_DIR_OVERRIDE=$(cfg_get_value "repository" "repos_cfg_dir")
+export EESSI_REPOS_CFG_DIR_OVERRIDE=${EESSI_REPOS_CFG_DIR_OVERRIDE:-${PWD}/cfg}
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh:98]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+SINGULARITY_CACHEDIR=$(cfg_get_value "site_config" "container_cachedir")
+echo "SINGULARITY_CACHEDIR='${SINGULARITY_CACHEDIR}'"
+if [[ ! -z ${SINGULARITY_CACHEDIR} ]]; then
+    export SINGULARITY_CACHEDIR
+fi
+
+echo -n "setting \$STORAGE by replacing any var in '${LOCAL_TMP}' -> "
+# replace any env variable in ${LOCAL_TMP} with its
+#   current value (e.g., a value that is local to the job)
+STORAGE=$(envsubst <<< ${LOCAL_TMP})
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh:64]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+echo "'${STORAGE}'"
+
+# obtain list of modules to be loaded
+LOAD_MODULES=$(cfg_get_value "site_config" "load_modules")
+echo "LOAD_MODULES='${LOAD_MODULES}'"
+
+# singularity/apptainer settings: CONTAINER, HOME, TMPDIR, BIND
+CONTAINER=$(cfg_get_value "repository" "container")
+export SINGULARITY_HOME="$(pwd):/eessi_bot_job"
+export SINGULARITY_TMPDIR="$(pwd)/singularity_tmpdir"
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh]: Left essentially as it was, but changed `$(pwd)` to `${PWD}`

File: bot/build.sh [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+STORAGE=$(envsubst <<< ${LOCAL_TMP})
+echo "'${STORAGE}'"
+
+# obtain list of modules to be loaded
+LOAD_MODULES=$(cfg_get_value "site_config" "load_modules")
+echo "LOAD_MODULES='${LOAD_MODULES}'"
+
+# singularity/apptainer settings: CONTAINER, HOME, TMPDIR, BIND
+CONTAINER=$(cfg_get_value "repository" "container")
+export SINGULARITY_HOME="$(pwd):/eessi_bot_job"
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+source scripts/cfg_files.sh
+
+# check if './cfg/job.cfg' exists
+if [[ ! -r "${JOB_CFG_FILE}" ]]; then
+    fatal_error "job config file (JOB_CFG_FILE=${JOB_CFG_FILE}) does not exist or not readable"
+fi
+echo "obtaining configuration settings from '${JOB_CFG_FILE}'"
+cfg_load ${JOB_CFG_FILE}
+
+# if http_proxy is in cfg/job.cfg use it, if not use env var $http_proxy
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+echo "bot/build.sh: Showing job.cfg from software-layer side"
+cat cfg/job.cfg
+
+# source utils.sh and cfg_files.sh
+source scripts/utils.sh
+source scripts/cfg_files.sh
+
+# check if './cfg/job.cfg' exists
+if [[ ! -r "${JOB_CFG_FILE}" ]]; then
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh:32]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+# defaults
+export JOB_CFG_FILE="${JOB_CFG_FILE_OVERRIDE:=./cfg/job.cfg}"
+
+echo "bot/build.sh: Showing job.cfg from software-layer side"
+cat cfg/job.cfg
+
+# source utils.sh and cfg_files.sh
+source scripts/utils.sh
+source scripts/cfg_files.sh
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh:26]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: bot/build.sh [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+#    pull request (OR by some other means)
+#  - the working directory contains a directory 'cfg' where the main config
+#    file 'job.cfg' has been deposited
+#  - the directory may contain any additional files references in job.cfg
+
+# defaults
+export JOB_CFG_FILE="${JOB_CFG_FILE_OVERRIDE:=./cfg/job.cfg}"
+
+echo "bot/build.sh: Showing job.cfg from software-layer side"
+cat cfg/job.cfg
```
[the remark this answers is not in the document]
Comment (trz42) [reply bot/build.sh]: Addressed in https://github.com/EESSI/software-layer/pull/233/commits/d08bd211ed6c23c5b917b01bd01ef75a5cb7b1b9

File: eessi_container.sh (hunk 1 above)
Comment (trz42) [reply eessi_container.sh:178]: See [PR#103](https://github.com/EESSI/docs/pull/103)

File: bot/build.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (boegel) [reply bot/build.sh:64]: see `$JOB_STORAGE` below

File: bot/build.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (boegel) [reply bot/build.sh]: OK

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Merging this, since it's ready, and so is https://github.com/EESSI/eessi-bot-software-layer/pull/155

**Pull request merged**